### PR TITLE
change case defaults

### LINKED ISSUE #803: UPPER CASE, lower case and Title Case
In Rust, we typically use `TitleCase` for types (like `struct Uart0 { .. .}`), and `lower_case` for module names (like `pac::uart0`), method/function names (like `my_uart.baudrate()`), and field names within structs (like `my_uart.baudrate`). Things that are `UPPER_CASE` are typically only static variables and constants.

The code generated by svd2rust currently uses a lot of `UPPER_CASE` names for types, and also for the field names in `struct Peripherals`. I think this can be confusing for a newcomer to Embedded Rust.

* Does anyone recall the reason for using `UPPER_CASE` here?
* Is there any appetite for changing the output to follow a more typical approach for capitalising things?

### PR DESCRIPTION
**Pull request by burrbull**
Closes #803 

cc @jonathanpallant, @Emilgardis

### CODE REVIEW

**Comment by Emilgardis** [comment]
I think it would be worth with this to add `#[doc(alias = "<peripheral name in svd>")]`

**Comment by burrbull** [comment]
Looks like starting `_` sometimes incompatible with `PascalCase`.

**Comment by burrbull** [comment]
I don't have ideas what to do with Freescale.

**Comment by Emilgardis** [comment]
ugh, why does the /ci triggers fail here, I have no issues with https://github.com/cross-rs/cross/blob/main/.github/workflows/try.yml ...

/ci diff pr

**Comment by burrbull** [comment]
> ugh, why does the /ci triggers fail here, I have no issues with

Are you sure someone else except you can run this?

**Comment by Emilgardis** [comment]
There's nothing in it that would prohibit other members to run it, the only check is that the user making the comment is a member, which we both are

**Comment by Emilgardis** [comment]
```
    {
      "id": "IC_kwDOBDHJwM5yw0Sf",
      "author": {
        "login": "burrbull"
      },
      "authorAssociation": "MEMBER",
```

I don't understand... 

/ci diff pr

**Comment by burrbull** [comment]
@Emilgardis run ci yet one time, please.

**Comment by Emilgardis** [comment]
can you try it, this time put some text before

**Comment by Emilgardis** [comment]
/ci diff --pager "git diff --no-index" pr

**Comment by Emilgardis** [comment]
original behaviour would be

```

--ident-format field_reader::c:_R 
--ident-format field_writer::c:_W 
--ident-format enum_name::c:_A 
--ident-format enum_write_name::c:_AW 
--ident-format enum_value::c: 
--ident-format interrupt::c: 
--ident-format cluster::c: 
--ident-format cluster_accessor::c: 
--ident-format cluster_mod::s:
--ident-format register::c: 
--ident-format register_spec::c:_SPEC 
--ident-format register_accessor::c: 
--ident-format register_mod::s:
--ident-format peripheral::c:
--ident-format peripheral_sigleton::c:
--ident-format peripheral_mod::s:
--ident-format peripheral_feature::c:
```

File: src/config.rs [hunk 1]
```diff
@@ -169,32 +174,66 @@ impl IdentFormat {
...
+            IdentFormat::default().pascal_case().suffix("Spec"),
+        );
+        map.insert(
+            "register_accessor".into(),
+            IdentFormat::default().snake_case(),
+        );
+        map.insert("register_mod".into(), IdentFormat::default().snake_case());
+        map.insert("peripheral".into(), IdentFormat::default().pascal_case());
+        map.insert(
+            "peripheral_sigleton".into(),
```
Comment (Emilgardis) [inline src/config.rs:216]: typo

also in device.rs L240 and L257

**Comment by burrbull** [comment]
Seems like that's all

**Comment by burrbull** [comment]
Lets try

/ci diff pr

**Comment by Emilgardis** [comment]
I don't understand why you're not able to trigger it. Might just have to switch to a check in the binary instead of relying on githubs actions environment doing things correctly...

/ci diff pr

/ci diff --current "@pr --ident-format field_reader::c:_R --ident-format field_writer::c:_W --ident-format enum_name::c:_A --ident-format enum_write_name::c:_AW --ident-format enum_value::c: --ident-format interrupt::c: --ident-format cluster::c: --ident-format cluster_accessor::c: --ident-format cluster_mod::s: --ident-format register::c: --ident-format register_spec::c:_SPEC --ident-format register_accessor::c: --ident-format register_mod::s: --ident-format peripheral::c: --ident-format peripheral_singleton::c: --ident-format peripheral_mod::s: --ident-format peripheral_feature::c:"

**Comment by Emilgardis** [comment]
/ci diff pr

/ci diff --current "@pr --ident-format field_reader::c:_R --ident-format field_writer::c:_W --ident-format enum_name::c:_A --ident-format enum_write_name::c:_AW --ident-format enum_value::c: --ident-format interrupt::c: --ident-format cluster::c: --ident-format cluster_accessor::c: --ident-format cluster_mod::s: --ident-format register::c: --ident-format register_spec::c:_SPEC --ident-format register_accessor::c: --ident-format register_mod::s: --ident-format peripheral::c: --ident-format peripheral_singleton::c: --ident-format peripheral_mod::s: --ident-format peripheral_feature::c:"

**Comment by Emilgardis** [comment]
/ci diff -c stm32f103 --current "@pr --ident-format field_reader::c:_R --ident-format field_writer::c:_W --ident-format enum_name::c:_A --ident-format enum_write_name::c:_AW --ident-format enum_value::c: --ident-format interrupt::c: --ident-format cluster::c: --ident-format cluster_accessor::c: --ident-format cluster_mod::s: --ident-format register::c: --ident-format register_spec::c:_SPEC --ident-format register_accessor::c: --ident-format register_mod::s: --ident-format peripheral::c: --ident-format peripheral_singleton::c: --ident-format peripheral_mod::s: --ident-format peripheral_feature::c:"

**Comment by Emilgardis** [comment]
oops, actual should be 

```
--ident-format field_reader::c:_R                                                                                                                          
--ident-format field_writer::c:_W 
--ident-format enum_name::c:_A 
--ident-format enum_write_name::c:_AW 
--ident-format enum_value::c: 
--ident-format interrupt::c: 
--ident-format cluster::c: 
--ident-format cluster_accessor::s: 
--ident-format cluster_mod::s:
--ident-format register::c: 
--ident-format register_spec::c:_SPEC 
--ident-format register_accessor::s: 
--ident-format register_mod::s:
--ident-format peripheral::c:
--ident-format peripheral_singleton::c:
--ident-format peripheral_mod::s:
--ident-format peripheral_feature::c:
```

/ci diff -c stm32f103 --current "@pr --ident-format field_reader::c:_R --ident-format field_writer::c:_W --ident-format enum_name::c:_A --ident-format enum_write_name::c:_AW --ident-format enum_value::c: --ident-format interrupt::c: --ident-format cluster::c: --ident-format cluster_accessor::s: --ident-format cluster_mod::s: --ident-format register::c: --ident-format register_spec::c:_SPEC --ident-format register_accessor::s: --ident-format register_mod::s: --ident-format peripheral::c: --ident-format peripheral_singleton::c: --ident-format peripheral_mod::s: --ident-format peripheral_feature::c:"

File: src/main.rs [hunk 2]
```diff
@@ -29,7 +32,35 @@ fn parse_configs(app: Command) -> Result<Config> {
...
+            if let (Some(n), Some(p), Some(c), Some(s)) = (f.next(), f.next(), f.next(), f.next()) {
+                let case = match c {
+                    "" | "unchanged" | "svd" => None,
+                    "p" | "pascal" | "type" => Some(Case::Pascal),
+                    "s" | "snake" | "lower" => Some(Case::Snake),
+                    "c" | "constant" | "upper" => Some(Case::Constant),
+                    _ => return Err(anyhow!("Unknown case")),
+                };
+                config.ident_formats.insert(
+                    n.into(),
```
[the remark this answers is not in the document]
Comment (Emilgardis) [reply src/main.rs:52]: I don't think it has to be a hard error that quits the program early

**Review by Emilgardis — CHANGES_REQUESTED** [review]
LGTM! Maybe we should document how to revert this behaviour more explicitly

**Comment by burrbull** [comment]
cc @Emilgardis

File: src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Emilgardis) [reply src/main.rs:55]: this is not solved, an invalid name, e.g `cluster_accesor` doesn't warn

**Review by Emilgardis — CHANGES_REQUESTED** [review]
The way to revert this behaviour would be to use a config containing

```toml
#svd2rust.toml
[ident_formats.field_reader]
case = "constant"
suffix = "_R"
[ident_formats.field_writer]
case = "constant"
suffix = "_W"
[ident_formats.enum_name]
case = "constant"
suffix = "_A"
[ident_formats.enum_write_name]
case = "constant"
suffix = "_AW"
[ident_formats.enum_value]
case = "constant"
[ident_formats.interrupt]
case = "constant"
[ident_formats.cluster]
case = "constant"
[ident_formats.cluster_accessor]
case = "snake"
[ident_formats.cluster_mod]
case = "snake"
[ident_formats.register]
case = "constant"
[ident_formats.register_spec]
case = "constant"
suffix = "_SPEC"
[ident_formats.register_accessor]
case = "snake"
[ident_formats.register_mod]
case = "snake"
[ident_formats.peripheral]
case = "constant"
[ident_formats.peripheral_singleton]
case = "constant"
[ident_formats.peripheral_mod]
case = "snake"
[ident_formats.peripheral_feature]
case = "constant"
```

I'm not sure myself if that is the best way to go about it, maybe we could have "themes" where there's a default theme (the new behaviour, and another theme that looks like how it used to look like. You can still set individual formats if wanted.

so maybe

```toml
ident_formats_theme = "legacy"
```

I also have a concern about the validation

**Comment by BrokenR3C0RD** [comment]
Tested this on a PAC I'm working on for the LPC11U1x/2x/3x MCUs, and it leads to some bad names for pins. In the SVD and reference manual, they all follow the naming scheme of `PIO{port}_{pin}`, where `port` is `0,1`, and `pin` is `0..=31`. However, the defaults used in this branch result in `Pio{port}{pin}`, resulting in pin types being named things like `Pio19`, `Pio110`, etc. Additionally, some of the `enumeratedValues` names get treated poorly too; one example is `BETWEEN_1_20MHZ`, which turns into `Between120Mhz`, and `BETWEEN_15_20MHZ`, which turns into `Between1520Mhz`.

Personal opinion, but I believe defaults should keep underscores between numbers; `Between15_20Mhz` isn't the prettiest, but would be more sensible than the current defaults.

**Comment by burrbull** [comment]
> Tested this on a PAC I'm working on for the LPC11U1x/2x/3x MCUs, and it leads to some bad names for pins. In the SVD and reference manual, they all follow the naming scheme of `PIO{port}_{pin}`, where `port` is `0,1`, and `pin` is `0..=31`. However, the defaults used in this branch result in `Pio{port}{pin}`, resulting in pin types being named things like `Pio19`, `Pio110`, etc. Additionally, some of the `enumeratedValues` names get treated poorly too; one example is `BETWEEN_1_20MHZ`, which turns into `Between120Mhz`, and `BETWEEN_15_20MHZ`, which turns into `Between1520Mhz`.
> 
> Personal opinion, but I believe defaults should keep underscores between numbers; `Between15_20Mhz` isn't the prettiest, but would be more sensible than the current defaults being used in this branch.

It looks like we need custom logic for `to_pascal_case` instead of built-in. I've tried to make some workarounds but those fix only individual cases.

**Comment by burrbull** [comment]
@BrokenR3C0RD try last changes

**Comment by BrokenR3C0RD** [comment]
@burrbull Works as expected, `PIO0_11` becomes `Pio0_11` and `BETWEEN_15_20_MHZ` becomes `Between15_20Mhz`. However, I think there might be an issue with 324b12c (which stops the panic in #811). [SVD](https://gist.github.com/BrokenR3C0RD/9fa8a1697b5bca9c1ab81221c33a688f/074fd8894c4d957bb96ec3659bbaf0c2893f9768)

Generates the following for PIO1:
```rs
// ...
impl RegisterBlock {
    #[doc = "0x60..0xe0 - Generic I/O configuration for PIO pins, port 1."]
    #[inline(always)]
    pub const fn pio1(&self, n: usize) -> &Pio10 {
        #[allow(clippy::no_effect)]
        [(); 32][n];
        unsafe { &*(self as *const Self).cast::<u8>().add(96).add(4 * n).cast() }
    }
    #[doc = "Iterator for array of:"]
    #[doc = "0x60..0xe0 - Generic I/O configuration for PIO pins, port 1."]
    #[inline(always)]
    pub fn pio1_iter(&self) -> impl Iterator<Item = &Pio10> {
        (0..32)
            .map(move |n| unsafe { &*(self as *const Self).cast::<u8>().add(96).add(4 * n).cast() })
    }
}

#[doc = "PIO0 (rw) register accessor: Generic I/O configuration for PIO, port 0.\n\nYou can [`read`](crate::generic::Reg::read) this register and get [`pio0::R`].  You can [`write_with_zero`](crate::generic::Reg::write_with_zero) this register using [`pio0::W`]. You can also [`modify`](crate::generic::Reg::modify) this register. See [API](https://docs.rs/svd2rust/#read--modify--write-api).\n\nFor information about available fields see [`mod@pio0`]
module"]
#[doc(alias = "PIO0")]
pub type Pio0 = crate::Reg<pio0::Pio0Spec>;
#[doc = "Generic I/O configuration for PIO, port 0."]
pub mod pio0;
pub use pio0 as pio1;
pub use Pio0 as Pio1;
```

Note that `Pio10` is referenced instead of `Pio1`.

**Comment by burrbull** [comment]
> Note that Pio10 is referenced instead of Pio1.

I think this is due to https://github.com/rust-embedded/svd2rust/blob/a23a780b091b238299ad91abb267fb9a2cbadef2/src/generate/peripheral.rs#L1192. But I don't remember when and why it was added. Maybe this is not needed more.

**Comment by Emilgardis** [comment]
git blame reveals it was https://github.com/rust-embedded/svd2rust/pull/662

**Comment by BrokenR3C0RD** [comment]
a085e4e now generates the correct `pio1`/`pio1_iter` functions for me, so #811 can be closed when/if this is merged.

**Comment by burrbull** [comment]
> this is not solved, an invalid name, e.g cluster_accesor doesn't warn

Fixed.

@Emilgardis can we discuss themes in separate PR?

**Comment by Emilgardis** [comment]
> @Emilgardis can we discuss themes in separate PR?

Sounds good!

**Review by Emilgardis — APPROVED** [review]
LGTM

**Comment by Emilgardis** [comment]
Should we discuss this broader or are we good with going forward with this breaking change? I kinda want one more team member to ack

**Comment by burrbull** [comment]
> Should we discuss this broader or are we good with going forward with this breaking change? I kinda want one more team member to ack

I can split of default formats changing (maybe with themes) in separate PR and we discuss them on next meeting.

**Comment by Emilgardis** [comment]
Lets keep the default changes to this pr but split out the ident-format improvements to a new pr

**Comment by burrbull** [comment]
> Lets keep the default changes to this pr but split out the ident-format improvements to a new pr which we can then just merge

#812

**Comment by burrbull** [comment]
> I'm not sure myself if that is the best way to go about it, maybe we could have "themes" where there's a default theme (the new behaviour), and another theme that looks like how it used to look like. You can still set individual formats if wanted.
> so maybe
> ```toml
> ident_formats_theme = "legacy"
> ```

Done. Not sure about `new` theme name.

File: src/config.rs [hunk 3]
```diff
@@ -172,65 +173,113 @@ impl IdentFormat {
...
+            ("field_accessor".into(), IdentFormat::default().snake_case()),
+            (
+                "field_reader".into(),
+                IdentFormat::default().pascal_case().suffix("R"),
+            ),
+            (
+                "field_writer".into(),
+                IdentFormat::default().pascal_case().suffix("W"),
+            ),
+            ("enum_name".into(), IdentFormat::default().pascal_case()),
```
Comment (Emilgardis) [inline src/config.rs:192]: I'm not a fan of the verbosity of the code here. Also quite a bit of code duplication. Is there a way to minimize it?

**Comment by Emilgardis** [comment]
> Done. Not sure about `new` theme name.

We could also not have a `new` theme, e.g not expose it

**Comment by Emilgardis** [comment]
maybe `default`?

With this variant though (`legacy` and `default`/`new`), we get into a issue of versioning, where if we change the theme we'd probably have to introduce a new theme for the previous version. I think only having `legacy` makes sense for now. Maybe there's an argument for having ranged theme names, such that saying `theme = "0.35.1"` means you get the appropriate theme for that version, but sounds like hell to maintain and keep track of

**Comment by thejpster** [comment]
2021 and 2024?

File: src/config.rs (hunk 3 above)
Comment (burrbull) [reply src/config.rs]: Is now better?

Comment (Emilgardis) [reply src/config.rs]: yep, I like it much better now

**Comment by burrbull** [comment]
@Emilgardis Can you fix `svd2rust-regress` "unused" warnings?

**Review by adamgreig — APPROVED** [review]

**Review by Emilgardis — APPROVED** [review]

**Pull request merged**